src/MediaWiki/Search/Search.php

```diff
@@ -471,4 +446,52 @@ public function setShowSuggestion( $showSuggestion ) {
 		parent::setShowSuggestion( $showSuggestion );
 		$this->getFallbackSearchEngine()->setShowSuggestion( $showSuggestion );
 	}
+
+	/**
+	 * @return SearchEngine
+	 */
+	private function newFallbackSearchEngine() {
+
+		$applicationFactory = ApplicationFactory::getInstance();
+		$type = $applicationFactory->getSettings()->get( 'smwgFallbackSearchType' );
+
+		$connection = $this->getDB();
+
+		if ( is_callable( $type ) ) {
+			// #3939
+			$fallbackSearch = $type( $connection );
+		} elseif ( $type !== null && $this->isValidFallbackSearchEngineType( $type ) ) {
+			$fallbackSearch = new $type( $connection );
```

Comment:
This could still lead to a crash.
Using a string for `$smwgFallbackSearchType` is still possible and `isValidFallbackSearchEngineType()` will allow any subclass of `SearchEngine`, including those where the constructor does not take a DB connection as parameter (e.g. CirrusSearch).
I see two ways: 1. Check for `!is_subclass_of( $type, 'SearchDatabase' )` in `isValidFallbackSearchEngineType()` or 2. just wrap the `new $type( $connection )` in a try-catch.

Reply:
Yes, so maybe:

- Check for `!is_subclass_of( $type, 'SearchDatabase' )` in isValidFallbackSearchEngineType()
- Rename `isValidFallbackSearchEngineType` to `isValidFallbackSearchDatabaseType`